something wrong in TypedSpecSpec

After the Deferred refactoring TypedSpecSpec is unhappy (but doesn't fail).

```
[ERROR] [04/06/2017 08:23:47.528] [TypedSpecSpec_A_u0020TypedSpec_CommonTests__anonfun_must_u0020report_u0020failures_1-akka.actor.default-dispatcher-2] [typed://TypedSpecSpec_A_u0020TypedSpec_CommonTests__anonfun_must_u0020report_u0020failures_1/user/failure] expected
org.scalatest.exceptions.TestFailedException: expected
	at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:528)
	at org.scalatest.refspec.RefSpec.newAssertionFailedException(RefSpec.scala:1331)
	at org.scalatest.Assertions$class.fail(Assertions.scala:1089)
	at org.scalatest.refspec.RefSpec.fail(RefSpec.scala:1331)
	at akka.typed.TypedSpecSpec$A$u0020TypedSpec$CommonTests$$anonfun$must$u0020report$u0020failures$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$2.apply(TypedSpec.scala:199)
	at akka.typed.TypedSpecSpec$A$u0020TypedSpec$CommonTests$$anonfun$must$u0020report$u0020failures$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$2.apply(TypedSpec.scala:199)
	at akka.typed.StepWise$.akka$typed$StepWise$$run(StepWise.scala:134)
	at akka.typed.StepWise$$anonfun$apply$1.apply(StepWise.scala:113)
	at akka.typed.StepWise$$anonfun$apply$1.apply(StepWise.scala:112)
	at akka.typed.scaladsl.Actor$Deferred.apply(Actor.scala:212)
	at akka.typed.Behavior$.undefer(Behavior.scala:159)
	at akka.typed.Behavior$.preStart(Behavior.scala:184)
	at akka.typed.internal.SupervisionMechanics$class.create(SupervisionMechanics.scala:73)
	at akka.typed.internal.SupervisionMechanics$class.processSignal(SupervisionMechanics.scala:54)
	at akka.typed.internal.ActorCell.processSignal(ActorCell.scala:68)
	at akka.typed.internal.ActorCell.liftedTree1$1(ActorCell.scala:384)
	at akka.typed.internal.ActorCell.processAllSystemMessages(ActorCell.scala:384)
	at akka.typed.internal.ActorCell.run(ActorCell.scala:281)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[ERROR] [04/06/2017 08:23:47.530] [TypedSpecSpec_A_u0020TypedSpec_CommonTests__anonfun_must_u0020report_u0020failures_1-akka.actor.default-dispatcher-2] [typed://TypedSpecSpec_A_u0020TypedSpec_CommonTests__anonfun_must_u0020report_u0020failures_1/user/failure] failure during PostStop
java.lang.IllegalArgumentException: deferred should not be passed to interpreter
	at akka.typed.Behavior$.interpret(Behavior.scala:214)
	at akka.typed.Behavior$.interpretSignal(Behavior.scala:209)
	at akka.typed.internal.SupervisionMechanics$class.finishTerminate(SupervisionMechanics.scala:93)
	at akka.typed.internal.ActorCell.finishTerminate(ActorCell.scala:68)
	at akka.typed.internal.SupervisionMechanics$class.terminate(SupervisionMechanics.scala:83)
	at akka.typed.internal.SupervisionMechanics$class.processSignal(SupervisionMechanics.scala:55)
	at akka.typed.internal.ActorCell.processSignal(ActorCell.scala:68)
	at akka.typed.internal.ActorCell.liftedTree1$1(ActorCell.scala:384)
	at akka.typed.internal.ActorCell.processAllSystemMessages(ActorCell.scala:384)
	at akka.typed.internal.ActorCell.run(ActorCell.scala:281)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

-> system LocalActorRef class akka.typed.scaladsl.Actor$Stateless status=0 nextMsg=null
   ⌊-> 1-receptionist LocalActorRef class akka.typed.scaladsl.Actor$Stateful status=0 nextMsg=null
   ⌊-> 2-eventStreamUnsubscriber LocalActorRef class akka.typed.scaladsl.Actor$Stateful status=0 nextMsg=null
   ⌊-> 3-log1-DefaultLogger LocalActorRef class akka.typed.Behavior$EmptyBehavior$ status=0 nextMsg=null
       ⌊-> logger LocalActorRef class akka.typed.scaladsl.Actor$Stateless status=0 nextMsg=null
-> user LocalActorRef class akka.typed.scaladsl.Actor$Stateful status=0 nextMsg=null
-> system LocalActorRef class akka.typed.scaladsl.Actor$Stateless status=0 nextMsg=null
   ⌊-> 1-receptionist LocalActorRef class akka.typed.scaladsl.Actor$Stateful status=0 nextMsg=null
   ⌊-> 2-eventStreamUnsubscriber LocalActorRef class akka.typed.scaladsl.Actor$Stateful status=0 nextMsg=null
   ⌊-> 3-log1-DefaultLogger LocalActorRef class akka.typed.Behavior$EmptyBehavior$ status=0 nextMsg=null
       ⌊-> logger LocalActorRef class akka.typed.scaladsl.Actor$Stateless status=0 nextMsg=null
-> user LocalActorRef class akka.typed.scaladsl.Actor$Stateful status=0 nextMsg=null
```

--- a/akka-typed-testkit/src/main/scala/akka/typed/testkit/Effects.scala
+++ b/akka-typed-testkit/src/main/scala/akka/typed/testkit/Effects.scala
@@ -51,7 +51,7 @@ class EffectfulActorContext[T](_name: String, _initialBehavior: Behavior[T], _ma
   }
   def hasEffects: Boolean = effectQueue.peek() != null
 
-  private var current = Behavior.preStart(_initialBehavior, this)
+  private var current = Behavior.validateAsInitial(Behavior.undefer(_initialBehavior, this))
 
   def currentBehavior: Behavior[T] = current
   def isAlive: Boolean = Behavior.isAlive(current)

--- a/akka-typed/src/main/scala/akka/typed/Behavior.scala
+++ b/akka-typed/src/main/scala/akka/typed/Behavior.scala
@@ -155,13 +155,9 @@ object Behavior {
     }
 
   @tailrec
-  def undefer[T](deferredBehavior: DeferredBehavior[T], ctx: ActorContext[T]): Behavior[T] = {
-    val behavior = deferredBehavior(ctx)
-
-    behavior match {
-      case innerDeferred: DeferredBehavior[T] @unchecked ⇒ undefer(innerDeferred, ctx)
-      case _ ⇒ behavior
-    }
+  def undefer[T](behavior: Behavior[T], ctx: ActorContext[T]): Behavior[T] = behavior match {
+    case innerDeferred: DeferredBehavior[T] @unchecked ⇒ undefer(innerDeferred(ctx), ctx)
+    case _ ⇒ behavior
   }
 
   /**
@@ -169,7 +165,7 @@ object Behavior {
    * notably the behavior can neither be `Same` nor `Unhandled`. Starting
    * out with a `Stopped` behavior is allowed, though.
    */
-  def validateAsInitial[T](behavior: Behavior[T]): Behavior[T] =
+  def validateAsInitial[T](behavior: BehaSo I get updates every day or so. Bleeding edge, but also works whenever I need it, which is quite rare as well. vior[T]): Behavior[T] =
     behavior match {
       case SameBehavior | UnhandledBehavior ⇒
         throw new IllegalArgumentException(s"cannot use $behavior as initial behavior")
@@ -179,12 +175,8 @@ object Behavior {
   /**
    * Validate the given behavior as initial, initialize it if it is deferred.
    */
-  def preStart[T](behavior: Behavior[T], ctx: ActorContext[T]): Behavior[T] = {
-    validateAsInitial(behavior) match {
-      case d: DeferredBehavior[T] @unchecked ⇒ validateAsInitial(undefer(d, ctx))
-      case x                                 ⇒ x
-    }
-  }
+  def preStart[T](behavior: Behavior[T], ctx: ActorContext[T]): Behavior[T] =
+    validateAsInitial(undefer(behavior, ctx))
 
   /**
    * Returns true if the given behavior is not stopped.
@@ -210,8 +202,8 @@ object Behavior {
 
   private def interpret[T](behavior: Behavior[T], ctx: ActorContext[T], msg: Any): Behavior[T] =
     behavior match {
-      case SameBehavior | UnhandledBehavior ⇒ throw new IllegalArgumentException(s"cannot execute with $behavior as behavior")
-      case _: DeferredBehavior[_]           ⇒ throw new IllegalArgumentException(s"deferred should not be passed to interpreter")
+      case SameBehavior | UnhandledBehavior ⇒ throw new IllegalArgumentException(s"cannot execute with [$behavior] as behavior")
+      case d: DeferredBehavior[_]           ⇒ throw new IllegalArgumentException(s"deferred [$d] should not be passed to interpreter")
       case IgnoreBehavior                   ⇒ SameBehavior.asInstanceOf[Behavior[T]]
       case StoppedBehavior                  ⇒ StoppedBehavior.asInstanceOf[Behavior[T]]
       case EmptyBehavior                    ⇒ UnhandledBehavior.asInstanceOf[Behavior[T]]
@@ -220,10 +212,7 @@ object Behavior {
           case signal: Signal ⇒ ext.management(ctx, signal)
           case msg            ⇒ ext.message(ctx, msg.asInstanceOf[T])
         }
-        possiblyDeferredResult match {
-          case d: DeferredBehavior[T] @unchecked ⇒ undefer(d, ctx)
-          case notDeferred                       ⇒ notDeferred
-        }
+        undefer(possiblyDeferredResult, ctx)
     }
 
 }

--- a/akka-typed/src/main/scala/akka/typed/Behavior.scala
+++ b/akka-typed/src/main/scala/akka/typed/Behavior.scala
@@ -165,18 +165,12 @@ object Behavior {
    * notably the behavior can neither be `Same` nor `Unhandled`. Starting
    * out with a `Stopped` behavior is allowed, though.
    */
-  def validateAsInitial[T](behavior: BehaSo I get updates every day or so. Bleeding edge, but also works whenever I need it, which is quite rare as well. vior[T]): Behavior[T] =
+  def validateAsInitial[T](behavior: Behavior[T]): Behavior[T] =
     behavior match {
       case SameBehavior | UnhandledBehavior ⇒
         throw new IllegalArgumentException(s"cannot use $behavior as initial behavior")
       case x ⇒ x
     }
-
-  /**
-   * Validate the given behavior as initial, initialize it if it is deferred.
-   */
-  def preStart[T](behavior: Behavior[T], ctx: ActorContext[T]): Behavior[T] =
-    validateAsInitial(undefer(behavior, ctx))
 
   /**
    * Returns true if the given behavior is not stopped.

--- a/akka-typed/src/main/scala/akka/typed/internal/SupervisionMechanics.scala
+++ b/akka-typed/src/main/scala/akka/typed/internal/SupervisionMechanics.scala
@@ -4,9 +4,7 @@
 package akka.typed
 package internal
 
-import scala.annotation.{ switch, tailrec }
 import scala.util.control.NonFatal
-import scala.util.control.Exception.Catcher
 import akka.event.Logging
 import akka.typed.Behavior.DeferredBehavior
 
@@ -89,8 +87,10 @@ private[typed] trait SupervisionMechanics[T] {
     val a = behavior
     /*
      * The following order is crucial for things to work properly. Only change this if you're very confident and lucky.
+     *
+     * Do not undefer a DeferredBehavior as that may cause creation side-effects, which we do not want on termination.
      */
-    try if (a ne null) Behavior.canonicalize(Behavior.interpretSignal(a, ctx, PostStop), a, ctx)
+    try if ((a ne null) && !a.isInstanceOf[DeferredBehavior[_]]) Behavior.canonicalize(Behavior.interpretSignal(a, ctx, PostStop), a, ctx)
     catch { case NonFatal(ex) ⇒ publish(Logging.Error(ex, self.path.toString, clazz(a), "failure during PostStop")) }
     finally try tellWatchersWeDied()
     finally try parent.sendSystem(DeathWatchNotification(self, failed))

--- a/akka-typed/src/main/scala/akka/typed/internal/SupervisionMechanics.scala
+++ b/akka-typed/src/main/scala/akka/typed/internal/SupervisionMechanics.scala
@@ -6,7 +6,7 @@ package internal
 
 import scala.util.control.NonFatal
 import akka.event.Logging
-import akka.typed.Behavior.DeferredBehavior
+import akka.typed.Behavior.{ DeferredBehavior, undefer, validateAsInitial }
 
 /**
  * INTERNAL API
@@ -68,7 +68,7 @@ private[typed] trait SupervisionMechanics[T] {
     behavior = initialBehavior
     if (system.settings.untyped.DebugLifecycle)
       publish(Logging.Debug(self.path.toString, clazz(behavior), "started"))
-    behavior = Behavior.preStart(behavior, ctx)
+    behavior = validateAsInitial(undefer(behavior, ctx))
     if (!Behavior.isAlive(behavior)) self.sendSystem(Terminate())
     true
   }

--- a/akka-typed/src/main/scala/akka/typed/internal/adapter/ActorAdapter.scala
+++ b/akka-typed/src/main/scala/akka/typed/internal/adapter/ActorAdapter.scala
@@ -58,11 +58,11 @@ import akka.annotation.InternalApi
   }
 
   override def preStart(): Unit =
-    behavior = Behavior.preStart(behavior, ctx)
+    behavior = validateAsInitial(undefer(behavior, ctx))
   override def preRestart(reason: Throwable, message: Option[Any]): Unit =
     next(Behavior.interpretSignal(behavior, ctx, PreRestart), PreRestart)
   override def postRestart(reason: Throwable): Unit =
-    behavior = Behavior.preStart(behavior, ctx)
+    behavior = validateAsInitial(undefer(behavior, ctx))
   override def postStop(): Unit = {
     next(Behavior.interpretSignal(behavior, ctx, PostStop), PostStop)
   }

--- a/akka-typed/src/main/scala/akka/typed/patterns/Restarter.scala
+++ b/akka-typed/src/main/scala/akka/typed/patterns/Restarter.scala
@@ -7,7 +7,7 @@ package patterns
 import scala.reflect.ClassTag
 import scala.util.control.NonFatal
 import akka.event.Logging
-import akka.typed.Behavior.{ DeferredBehavior, SameBehavior, StoppedBehavior, UnhandledBehavior }
+import akka.typed.Behavior._
 import akka.typed.scaladsl.Actor._
 
 /**
@@ -22,7 +22,7 @@ final case class Restarter[T, Thr <: Throwable: ClassTag](initialBehavior: Behav
   private def restart(ctx: ActorContext[T], startedBehavior: Behavior[T]): Behavior[T] = {
     try Behavior.interpretSignal(startedBehavior, ctx, PreRestart) catch { case NonFatal(_) ⇒ }
     // no need to canonicalize, it's done in the calling methods
-    Behavior.preStart(initialBehavior, ctx)
+    validateAsInitial(undefer(initialBehavior, ctx))
   }
 
   private def canonical(b: Behavior[T], ctx: ActorContext[T]): Behavior[T] =
@@ -78,13 +78,13 @@ final case class MutableRestarter[T, Thr <: Throwable: ClassTag](initialBehavior
 
   private[this] var current: Behavior[T] = _
   private def startCurrent(ctx: ActorContext[T]) = {
-    // we need to pre-start it the first time we have access to a context
-    if (current eq null) current = Behavior.preStart(initialBehavior, ctx)
+    // we need to undefer it if needed the first time we have access to a context
+    if (current eq null) current = validateAsInitial(undefer(initialBehavior, ctx))
   }
 
   private def restart(ctx: ActorContext[T]): Behavior[T] = {
     try Behavior.interpretSignal(current, ctx, PreRestart) catch { case NonFatal(_) ⇒ }
-    Behavior.preStart(initialBehavior, ctx)
+    validateAsInitial(undefer(initialBehavior, ctx))
   }
 
   override def management(ctx: ActorContext[T], signal: Signal): Behavior[T] = {

--- a/akka-typed/src/main/scala/akka/typed/patterns/Restarter.scala
+++ b/akka-typed/src/main/scala/akka/typed/patterns/Restarter.scala
@@ -32,15 +32,13 @@ final case class Restarter[T, Thr <: Throwable: ClassTag](initialBehavior: Behav
     else if (b eq behavior) Same
     else {
       b match {
-        case d: DeferredBehavior[_] ⇒ canonical(Behavior.undefer(d.asInstanceOf[DeferredBehavior[T]], ctx), ctx)
+        case d: DeferredBehavior[_] ⇒ canonical(Behavior.undefer(d, ctx), ctx)
         case b                      ⇒ Restarter[T, Thr](initialBehavior, resume)(b)
       }
     }
 
-  private def preStart(b: Behavior[T], ctx: ActorContext[T]): Behavior[T] = b match {
-    case d: DeferredBehavior[_] ⇒ Behavior.undefer(d.asInstanceOf[DeferredBehavior[T]], ctx)
-    case b                      ⇒ b
-  }
+  private def preStart(b: Behavior[T], ctx: ActorContext[T]): Behavior[T] =
+    Behavior.undefer(b, ctx)
 
   override def management(ctx: ActorContext[T], signal: Signal): Behavior[T] = {
     val startedBehavior = preStart(behavior, ctx)

